### PR TITLE
[Model Validation] Finishing touches and documentation

### PR DESCRIPTION
**Pull request by jerrylian-db**
## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

Here are the changes included in this PR:
* Created failure message for missing `baseline_model` parameter and added corresponding tests.
* Built a way to handle the case when baseline metric values equals 0 and added corresponding tests.
* Added parameter checks to MetricThreshold class instantiation and added corresponding tests.
* Polished doc strings and comments.
* Added model validation documentation and examples.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

I verified locally that the doc strings are built as expected.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/evaluation/base.py [hunk 1]
```diff
@@ -1077,26 +1084,27 @@ def scatter_plot(eval_df, builtin_metrics, artifacts_dir):
                                              :py:class:`mlflow.models.MetricThreshold` used for
                                              model validation. Each metric name must either be the
                                              name of a builtin metric or the name of a custom
-                                             metric defined in the``custom_metrics`` parameter.
+                                             metric defined in the ``custom_metrics`` parameter.
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:1087]: @jerrylian-db Can we add some documentation to `mlflow.evaluate()` (outside of the params section) explaining how this API can be used for model validation?

https://github.com/mlflow/mlflow/pull/6419#discussion_r948655290

File: mlflow/models/evaluation/validation.py [hunk 2]
```diff
@@ -13,9 +13,9 @@ class MetricThreshold:
                         >= threshold to pass validation.
                       - Otherwise, the metric value has to be <= threshold to pass the validation.
 
-    :param min_absolute_change: (Optional) A positive floating point number representing the
-                                minimum absolute change required for candidate model to
-                                pass the comparison with the baseline model.
+    :param min_absolute_change: (Optional) A positive number representing the minimum absolute
+                                change required for candidate model to pass the comparison with
```
Comment (dbczumar) [inline mlflow/models/evaluation/validation.py:17]:
```suggestion
                                change required for candidate model to pass validation with
```

File: mlflow/models/evaluation/base.py (hunk 1 above)
Comment (jerrylian-db) [reply mlflow/models/evaluation/base.py:1087]: I've included that in this PR. See the `Models.rst` changes and also the `examples/evaluation/evaluate_with_model_validation.py`

File: docs/source/models.rst [hunk 3]
```diff
@@ -1262,6 +1262,69 @@ dictionary of metrics, or two dictionaries representing metrics and artifacts.
...
 
+Performing Model Validation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can also use the :py:func:`mlflow.evaluate()` API to perform some checks on the metrics
+generated during model evaluation to validate the quality of your model. By specifying a 
+``validation_thresholds`` dictionary mapping metric names to :py:class:`mlflow.models.MetricThreshold` 
+objects, you can specify value thresholds that your model's evaluation metrics must exceed as well 
+as absolute and relative gains your model must have in comparison to a specified
+``baseline_model``. If you model fails to clear specified thresholds, :py:func:`mlflow.evaluate()`
```
Comment (dbczumar) [inline docs/source/models.rst:1273]:
```suggestion
``baseline_model``. If your model fails to clear specified thresholds, :py:func:`mlflow.evaluate()` 
```

File: docs/source/models.rst [hunk 4]
```diff
@@ -1262,6 +1262,69 @@ dictionary of metrics, or two dictionaries representing metrics and artifacts.
...
+Performing Model Validation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can also use the :py:func:`mlflow.evaluate()` API to perform some checks on the metrics
+generated during model evaluation to validate the quality of your model. By specifying a 
+``validation_thresholds`` dictionary mapping metric names to :py:class:`mlflow.models.MetricThreshold` 
+objects, you can specify value thresholds that your model's evaluation metrics must exceed as well 
+as absolute and relative gains your model must have in comparison to a specified
+``baseline_model``. If you model fails to clear specified thresholds, :py:func:`mlflow.evaluate()` 
+will throw a ``ModelValidationFailedException`` detailing why.
```
Comment (dbczumar) [inline docs/source/models.rst:1274]:
```suggestion
will throw a ``ModelValidationFailedException`` detailing the validation failure.
```

File: docs/source/models.rst [hunk 5]
```diff
@@ -1262,6 +1262,69 @@ dictionary of metrics, or two dictionaries representing metrics and artifacts.
...
+            eval_data,
+            targets="label",
+            model_type="classifier",
+            dataset_name="adult",
+            validation_thresholds=thresholds,
+            baseline_model=baseline_model_uri,
+        )
+
+Refer to :py:class:`mlflow.models.MetricThreshold` to see details on how the thresholds are specified
+and checked. For a more comprehensive demonstration on how to use perform model validation, refer to
```
Comment (dbczumar) [inline docs/source/models.rst:1324]:
```suggestion
and checked. For a more comprehensive demonstration on how to use :py:func:`mlflow.evaluate()` to perform model validation, refer to 
```

File: examples/evaluation/README.md [hunk 6]
```diff
@@ -1,7 +1,10 @@
 ### MLflow evaluation Examples
 
-The examples in this directory illustrate how you can use the `mlflow.evaluate` API to evaluate a PyFunc model on the
-specified dataset using builtin default evaluator, and log resulting metrics & artifacts to MLflow Tracking.
+The examples in this directory illustrate how you can use the `mlflow.evaluate` API. Specifically,
```
Comment (dbczumar) [inline examples/evaluation/README.md:3]:
```suggestion
The examples in this directory demonstrate how to use the `mlflow.evaluate()` API. Specifically,
```

File: examples/evaluation/evaluate_with_model_validation.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import xgboost
```
Comment (dbczumar) [inline examples/evaluation/evaluate_with_model_validation.py:1]: Nice! Can we add this to https://github.com/mlflow/mlflow/blob/b97155d7c4f210d63eae6588350068a7dc3f2db5/tests/examples/test_examples.py#L194?

File: mlflow/models/evaluation/validation.py [hunk 8]
```diff
@@ -45,55 +51,64 @@ def __init__(
...
-        Whether higher value is better for the metric.
-        :rtype: bool
+        Boolean value representing whether higher value is better for the metric.
         """
         return self._higher_is_better
 
     def is_empty(self):
         """
-        Return True if there is no threshold specified, False otherwise.
+        Returns True if there is no threshold specified, False otherwise.
```
Comment (dbczumar) [inline mlflow/models/evaluation/validation.py:111]: Nit: We should use pydoc `:return:` syntax here

**Review by dbczumar — APPROVED** [review]
LGTM with a few small comments. Thanks @jerrylian-db!

**Pull request merged**